autoretrieve/engine.go

```diff
@@ -0,0 +1,524 @@
+		return nil, nil
+	case DataTransferPublisher:
+		dtOpts := []dtsync.Option{dtsync.Topic(e.pubTopic), dtsync.WithExtraData(e.pubExtraGossipData)}
+		if e.pubDT != nil {
+			return dtsync.NewPublisherFromExisting(e.pubDT, e.h, e.pubTopicName, e.lsys, dtOpts...)
+		}
+		ds := dsn.Wrap(e.ds, datastore.NewKey("/legs/dtsync/pub"))
+		return dtsync.NewPublisher(e.h, ds, e.lsys, e.pubTopicName, dtOpts...)
+	case HttpPublisher:
+		return httpsync.NewPublisher(e.pubHttpListenAddr, e.lsys, e.h.ID(), e.key)
+	default:
+		return nil, fmt.Errorf("unknown publisher kind: %s", e.pubKind)
+	}
+}
+
+// PublishLocal stores the advertisement in the local link system and marks it locally as the latest
+// advertisement.
+//
+// The context is used for storing internal mapping information onto the datastore.
+//
+// See: AutoretrieveEngine.Publish.
+func (e *AutoretrieveEngine) PublishLocal(ctx context.Context, adv schema.Advertisement) (cid.Cid, error) {
+
+	if err := adv.Validate(); err != nil {
+		return cid.Undef, err
+	}
+
+	adNode, err := adv.ToNode()
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	lnk, err := e.lsys.Store(ipld.LinkContext{Ctx: ctx}, schema.Linkproto, adNode)
+	if err != nil {
+		return cid.Undef, fmt.Errorf("cannot generate advertisement link: %s", err)
+	}
+	c := lnk.(cidlink.Link).Cid
+	log := log.With("adCid", c)
+	log.Info("Stored ad in local link system")
+
+	if err := e.putLatestAdv(ctx, c.Bytes()); err != nil {
+		log.Errorw("Failed to update reference to the latest advertisement", "err", err)
+		return cid.Undef, fmt.Errorf("failed to update reference to latest advertisement: %w", err)
+	}
+	log.Info("Updated reference to the latest advertisement successfully")
+	return c, nil
+}
+
+// Publish stores the given advertisement locally via AutoretrieveEngine.PublishLocal first, then publishes
+// a message onto the gossipsub to signal the change in the latest advertisement by the provider to
+// indexer nodes.
+//
+// The publication mechanism uses legs.Publisher internally.
+// See: https://github.com/filecoin-project/go-legs
+func (e *AutoretrieveEngine) Publish(ctx context.Context, adv schema.Advertisement) (cid.Cid, error) {
+	c, err := e.PublishLocal(ctx, adv)
+	if err != nil {
+		log.Errorw("Failed to store advertisement locally", "err", err)
+		return cid.Undef, fmt.Errorf("failed to publish advertisement locally: %w", err)
+	}
+
+	// Only announce the advertisement CID if publisher is configured.
+	if e.publisher != nil {
+		log := log.With("adCid", c)
+		log.Info("Publishing advertisement in pubsub channel")
+		err = e.publisher.UpdateRoot(ctx, c)
+		if err != nil {
+			log.Errorw("Failed to announce advertisement on pubsub channel ", "err", err)
+			return cid.Undef, err
+		}
+	}
+	return c, nil
+}
+
+// PublishLatest re-publishes the latest existing advertisement to pubsub.
+func (e *AutoretrieveEngine) PublishLatest(ctx context.Context) error {
```

Comment:
how is this used? I can't see anywhere this is used

Reply:
I used it during testing but it turns out it wasn't the correct approach. I'll delete this